packages/template-retail-react-app/app/contexts/index.js

```diff
@@ -27,23 +29,157 @@ import PropTypes from 'prop-types'
+
+    const api = useCommerceAPI()
+    const [queue, setQueue] = useState({})
+    const [root, setRoot] = useState({
+        // map over the server-provided cat
+        ...treeRoot[DEFAULT_ROOT_CATEGORY],
+        [itemsKey]: treeRoot[DEFAULT_ROOT_CATEGORY]?.[itemsKey]?.map((item) => ({
+            ...item,
+            loaded: item?.loaded ?? false
+        }))
+    })
+
+    // iterates through each deep nested object and if finds object that has prop and value specified in objToFindBy
+    // argument, it replaces the current object with replacementObj, stops recursive walk and returns the whole tree.
+    // If none is found, it returns false.
+    const findAndModifyFirst = (tree, childrenKey, objToFindBy, replacementObj) => {
+        let treeToReturn = tree
+        let findSuccess = false
+        let modifiedObj = false
+        const findKeys = Object.keys(objToFindBy)
+        findKeys.forEach((key) => {
+            isEqual(tree[key], objToFindBy[key]) ? (findSuccess = true) : (findSuccess = false)
+        })
+        if (findSuccess) {
+            for (let prop in tree) {
+                delete tree[prop]
+            }
+            for (let prop in replacementObj) {
+                tree[prop] = replacementObj[prop]
+            }
+            return tree
+        }
+        const findInChildren = (obj, childrenKey, objToFindBy, replacementObj) => {
+            if (Object.prototype.hasOwnProperty.call(obj, childrenKey)) {
+                for (let i = 0; i < obj[childrenKey].length; i++) {
+                    findKeys.forEach((key) => {
+                        isEqual(obj[childrenKey][i][key], objToFindBy[key])
+                            ? (findSuccess = true)
+                            : (findSuccess = false)
+                    })
+                    if (findSuccess) {
+                        obj[childrenKey][i] = replacementObj
+                        modifiedObj = true
+                        break
+                    }
+                }
+                if (!findSuccess) {
+                    obj[childrenKey].forEach((child) =>
+                        findInChildren(child, childrenKey, objToFindBy, replacementObj)
+                    )
+                }
+            }
+            return obj
+        }
+        findInChildren(tree, childrenKey, objToFindBy, replacementObj)
+        return modifiedObj ? treeToReturn : false
+    }
+
+    const fetchCategoryNode = async (id, depth = 1) => {
+        const STALE_TIME = 10000 // 10 seconds
+        let res
+        // return early if there's an in-flight request or one that is less
+        // than the stale time
+        if (queue[id] === 'loading' || Date.now() < queue[id] + STALE_TIME) {
+            return
+        }
+        setQueue({
+            ...queue,
+            [id]: 'loading'
+        })
+        res = await api.shopperProducts.getCategory({
+            parameters: {
+                id,
+                levels: depth
```

Comment:
```suggestion
                levels
```
_Totally_ inconsequential, but you could change the variable `depths` to `levels` and then make this line real clean!